onnxruntime/python/onnxruntime_pybind_mlvalue.cc

```diff
@@ -612,20 +631,28 @@ void CreateGenericMLValue(const onnxruntime::InputDefList* input_def_list, const
 
     // The allocator will own the array memory and will decrement the reference on Free()
     // or when destroyed
-    auto pybind_allloc = std::make_shared<OrtPybindSingleUseAllocator>(arr, name_input, alloc->Info());
-    CreateTensorMLValueOwned(pybind_allloc, alloc, p_mlvalue);
-  } else if (PyList_Check(value.ptr())) {
+    auto pybind_alloc = std::make_shared<OrtPybindSingleUseAllocator>(arr, name_input, alloc->Info());
+    CreateTensorMLValueOwned(pybind_alloc, alloc, p_mlvalue);
+  } else if (!parse_numpy_only && PyList_Check(value.ptr())) {
     auto* seq_tensors = reinterpret_cast<PyObject*>(value.ptr());
     CreateSequenceOfTensors(alloc, name_input, input_def_list, seq_tensors, p_mlvalue);
-  } else if (PyDict_Check(value.ptr())) {
+  } else if (!parse_numpy_only && PyDict_Check(value.ptr())) {
 #if !defined(DISABLE_ML_OPS)
     CreateMapMLValue_AgnosticVectorMap((PyObject*)NULL, value.ptr(), alloc, name_input, p_mlvalue);
 #else
     ORT_UNUSED_PARAMETER(p_mlvalue);
     throw std::runtime_error("Map type is not supported in this build.");
 #endif
 
-  } else {
+  } else if (!parse_numpy_only && strcmp(Py_TYPE(value.ptr())->tp_name, "OrtValue") == 0) {
+    // This is an OrtValue coming in directly from Python, so assign the underlying native OrtValue handle
+    // to the OrtValue object that we are going to use for Run().
+    // This should just increase the ref counts of the underlying shared_ptrs in the native OrtValue
+    // and the ref count will be decreased when the OrtValue used for Run() is destroyed upon exit.
+    py::object native_ortvalue_handle = value.attr("_ortvalue");
+    auto native_ort_value = native_ortvalue_handle.cast<OrtValue>();
+    *p_mlvalue = native_ort_value;
+  } else if (!parse_numpy_only) {
```

Comment:
nit: does this incur an unnecessary copy? one to the native_ort_value and one to *p_mlvalue? Is it possible for the cast to return a pointer?

Reply:
Fixed